lib/pbench/server/filetree.py

```diff
@@ -336,72 +336,72 @@ def unpack(self, incoming: Path, results: Path):
             the tarball.
         """
         unpacked = incoming / f"{self.name}.unpack"
-        unpacked.mkdir(parents=True)
+        # unpacked.mkdir(parents=True)
 
         try:
             process = subprocess.run(
                 [
                     "tar",
-                    "-xf",
+                    "-x",
                     "--no-same-owner",
                     "--delay-directory-restore",
                     "--force-local",
-                    self.tarball_path,
+                    "--file=" + str(self.tarball_path),
                 ],
                 cwd=unpacked,
                 stdin=subprocess.DEVNULL,
                 capture_output=True,
                 text=True,
             )
         except Exception as exc:
-            raise TarballUnpackError(self.name, str(exc)) from exc
+            raise TarballUnpackError(self.tarball_path, str(exc)) from exc
         else:
             if process.returncode != 0:
                 raise TarballUnpackError(
-                    self.name,
+                    self.tarball_path,
                     f"tar exited with status {process.returncode}:  {process.stderr.strip()!r}",
                 )
 
-        command = "find . ( -type d -exec chmod ugo+rx '{}' + ) -o ( -type f -exec chmod ugo+r '{}' + )"
+        find_command = "find . ( -type d -exec chmod ugo+rx '{}' + ) -o ( -type f -exec chmod ugo+r '{}' + )"
         try:
             process = subprocess.run(
-                command.split(),
+                find_command.split(),
                 cwd=unpacked,
                 stdin=subprocess.DEVNULL,
                 capture_output=True,
                 text=True,
             )
         except Exception as exc:
-            raise TarballModeChangeError(self.name, str(exc)) from exc
+            raise TarballModeChangeError(self.tarball_path, str(exc)) from exc
         else:
             if process.returncode != 0:
                 raise TarballModeChangeError(
-                    self.name,
+                    self.tarball_path,
                     f"tar exited with status {process.returncode}:  {process.stderr.strip()!r}",
                 )
 
-        self.unpacked = incoming
+        self.unpacked = incoming / self.name
         unpack = unpacked / self.name
 
         try:
-            shutil.move(str(unpack), str(self.unpacked))
+            shutil.move(unpack, incoming)
         except Exception as exc:
-            raise TarballUnpackError(
-                self.name,
-                f"Error in moving unpacked tarball to Unpacked directory: {str(exc)}",
-            )
             try:
-                os.remove(self.unpacked / self.name)
+                shutil.rmtree(self.unpacked)
```

Comment:
I think here we want `unpacked`, which is the `${INCOMING}/{controller}/{name}.unpack` directory... we don't want to leave that behind.

Reply:
But here we are removing partially copied tarball in self.unpacked Directory, as the Tarball hasn't copied successfully it is raising error.

I checked for the `controller` they are joined with the incoming and results path and sent here. So, the incoming and results path that we have here already have controller attached with them.

Reply:
> I checked for the controller they are joined with the incoming and results path and sent here. So, the incoming and results path that we have here already have controller attached with them.

👍 

> Here we are removing partially copied tarball in self.unpacked Directory, as the Tarball hasn't copied successfully it is raising error.

It might be easier if we used the terms more precisely.  The "tarball" (in my mind, at least) is a single file whose name ends in `.tar` or `.tar.xz` or similar.  It contains an archive of a directory tree which we are trying to restore in a local filesystem.  So, at this point in the code, I believe that we've finished unpacking the tarball into a temporary directory, and the attempt to move that directory-tree into its final location has failed; and, so, we're trying to remove the temporary directory with the unpacked tarball contents in it.

The operation which failed is `shutil.move()`.  In the best case, this function removes the directory entry for its target from the target's parent directory and creates a new directory entry for the target at the specified destination.  Thus, there are no copy operations performed, and the operation is pretty close to atomic (i.e., it either works or it doesn't start).  However, there are circumstances where the function performs a copy-and-delete operation (e.g., if the source and destination are on different filesystems or devices).  In these cases, the operation might fail in the middle, but it should restore the situation back to the initial conditions (i.e., it should do all the copies first, and then the deletions, and, if any of the copies fail, it should remove the copies which it created).

So, when the failure occurs, we should need to remove the files and directories which the `tar` command created, but not much else.  We might need to remove the directory where the `tar` command placed the files which it restored from the tarball, and perhaps we need to remove the tarball itself.

But, what I think we're trying to do in the cited source lines is to undo the effect of the `tar` command.

Reply:
No, here we are just trying to undo the effect of `shutil.move` command.